sql/mysqld.cc

```diff
@@ -6418,38 +6450,200 @@ void handle_connections_sockets()
   int flags=0,retval;
   st_vio *vio_tmp;
   bool is_unix_sock;
-#ifdef HAVE_POLL
+  int systemd_n=0, systemd_fd;
+#ifdef HAVE_SYSTEMD
+  union sockaddr_union socketpeer;
+#endif
+  socklen_t socketpeer_len;
+  char const *family;
   int socket_count= 0;
-  struct pollfd fds[3]; // for ip_sock, unix_sock and extra_ip_sock
-  MYSQL_SOCKET  pfs_fds[3]; // for performance schema
-#define setup_fds(X)                    \
+  short fd_type[MAX_LISTEN_SOCKETS];
+#ifdef HAVE_POLL
+  struct pollfd fds[MAX_LISTEN_SOCKETS]; // for ip_sock, extra_ip_sock and unix_sock (could be any order for systemd)
+  MYSQL_SOCKET  pfs_fds[MAX_LISTEN_SOCKETS]; // for performance schema
+#define setup_fds(X,T)                    \
     mysql_socket_set_thread_owner(X);             \
     pfs_fds[socket_count]= (X);                   \
     fds[socket_count].fd= mysql_socket_getfd(X);  \
     fds[socket_count].events= POLLIN;   \
+    fd_type[socket_count]= T; \
     socket_count++
 #else
-#define setup_fds(X)    FD_SET(mysql_socket_getfd(X),&clientFDs)
+  int  pfs_flags[MAX_LISTEN_SOCKETS];
+  MYSQL_SOCKET  pfs_fds[MAX_LISTEN_SOCKETS]; // used for non-polling interfaces
+#define setup_fds(X,T)    FD_SET(mysql_socket_getfd(X),&clientFDs); \
+  pfs_fds[socket_count]= (X);                   \
+  fd_type[socket_count]= T; \
+  socket_count++
   fd_set readFDs,clientFDs;
   FD_ZERO(&clientFDs);
 #endif
 
   DBUG_ENTER("handle_connections_sockets");
 
-  if (mysql_socket_getfd(base_ip_sock) != INVALID_SOCKET)
+#ifdef HAVE_SYSTEMD
+  systemd_n = sd_listen_fds(1);
+  if (systemd_n < 0)
+  {
+    my_error(ER_SYSTEMD_LISTEN_FDS, MYF(0), 0, -systemd_n);
```

Comment:
ER_SYSTEMD_LISTEN_FDS has no additional parameters, while here you add "0, -systemd_n". Why?